### ISSUE TITLE
Model not able to learn with GT Pose from GPS/IMU

### ISSUE THREAD

**Issue by aj96** [nianticlabs/monodepth2#78]
I said I'd let you all know how it goes, so here it is. I tried training with GT cam2cam pose provided from GPS/IMU.

When trying to use these transformations to replace the PoseCNN, the model stops learning after about one epoch. And even more confusingly, the validation loss is consistently lower than the training loss. I even tried taking the high loss examples in the training set and using those as validation set instead, but the validation loss was still lower than training loss. 

The abs_relative error is about 0.15-0.16 with median scaling disabled, and the best model is from epoch 6 after training for 20 epochs. So the model is clearly not learning. The median scaling ratios are all very close to 1 though when I do enable median scaling. So it is learning the metric information. 

Here is the tensorboard:

<img width="1418" alt="tensorboard" src="https://user-images.githubusercontent.com/19962279/65558592-3e470280-deec-11e9-9480-d39865a0c918.png">

I am thinking that the validation loss being lower than the training loss is not too big of an issue because the loss function itself already has built-in erroneous assumptions. So it's not the best metric for whether the model is properly learning. We care about the evaluation metrics. But even for the evaluation metrics, it seems like validation is doing better.

I was wondering if these results make sense. I did notice that the model takes longer to learn reasonable disparity maps when taking away the PoseCNN, which makes sense. Because we've taken away half the architecture basically, halving its capacity to fit a model. Yet it just can't learn past the first epoch.

For reference, I explain my entire process:

I obtained the GT Camera-To-Camera Transformation by using this equation:

```
T_cam2cam =  T_velot_2_camt * T_imut_2_velot * T_enut0_2_imut * T_imut0_2_enut0 * T_velot0_2_imut0 * T_camt0_2_velot0
```

The only transformation that we don't have are the enu-to-imu transformations. I obtained this from the provided lat, lon, alt, roll (rx), pitch (ry), yaw (rz) at each timestep in the oxts folder. I used timestep zero as the reference local vertical frame for each sequence.  

```
import pymap3d as pm
tx, ty, tz = pm.geodetic2enu(lat_t, lon_t, alt_t, lat_t0, lon_t0, alt_t0)
```
tx, ty, tz form the translation vector t_imut_2_enut0

Now we need the rotation component R_imut_2_enut0. The rx, ry, rz only provide the rotation R_imut_2_enut. They are the rotations relative to local vertical. The complete rotation matrix is formed via the following: R_imut_2_enut0 = R_ecef_2_enut0 * R_enut_2_ecef * R_imut_2_enut.

That's how I obtained the T_imut_2_enut0 for every example. Of course, take the inverse to get T_enut0_2_imut.

Taking the point (0,0,0) at camera coordinate frame at time-step t0, I plotted the location of that point relative to every camera coordinate frame at time-step t and compared it to plotting the lat, lon points on google maps. Here's an example for 2011_09_26_drive_0001_sync/

<img width="393" alt="cam2cam_points_across_time" src="https://user-images.githubusercontent.com/19962279/65558268-e78cf900-deea-11e9-8a8f-1eccc93491f7.png">

<img width="1174" alt="2011_09_26_drive_001_waypoints" src="https://user-images.githubusercontent.com/19962279/65558275-ed82da00-deea-11e9-9604-57db3791f496.png">

As can be seen in the waypoints plotted on google maps, the car is moving forward and to the left, which means the point (0, 0, 0) relative to camera coordinate frame time-step 0 should be moving backwards and to the right. Z values are constantly decreasing. X values are constantly increasing in the 3d plot. There are some small fluctuations in Y towards the end, but they're small, less than a meter, so irrelevant. Total distance traveled is 107 meters which is about the distance between the first and last lat, lon points. Everything checks out.

**Comment by daniyar-niantic** [comment]
It is hard to say what the issue is, but could you try the following:
instead of feeding to the network the relative pose `(R, t)` that you computed, feed the inverse of the pose, i.e. `(np.linalg.inv(R), -np.dot(np.linalg.inv(R), t))`.

**Comment by aj96** [comment]
@daniyar-niantic Thanks for the suggestion. I had actually already tried that. I also tried inverting one of the matrices, but not the other. Any other combination causes the model to learn garbage. So I know that the matrix ordering is correct. I did notice that I forgot to disable the random horizontal flipping. That fixed the validation loss being lower than the training loss. But the model still doesn't achieve the same evaluation metrics as training with posecnn. 0.15 abs err vs 0.12 abs err for 128x416 resolution. Also, getting some infinite depth holes at texture-less regions. It does better with fine-tuning from a depth cnn that was trained with pose cnn. Abs rel err drops to 0.14. I am actually just realizing that I do not need to provide the entire camera to camera transformation. I can decompose it to: T_velot_2_camt * T_velot0_2_velot * T_camt0_2_velot0. I can provide the camera to velo transformation, which kitti provides, and just have the model learn the velo2velo transformation across time. Because the camera to velo transformation has the metric information, the depth should be metric as well, no? I wrote out the math, it seems like as long as rotation is not identity, there will be unique supervision provided to both the learned rotation and translation components for the velo 2 velo transformation.

**Comment by daniyar-niantic** [comment]
@aj96
- Your lower depth scores might be just due to the noise in the computed poses. IMU and GPS are not perfect.
- I'm not sure what are you trying to train, is it the depth network from KITTI poses, or are you training the pose prediction network? If it is just depth why do you need the model to learn the velo2velo transformation?
- I think if the relative poses between cameras are metric, the depth should be metric too.

**Comment by aj96** [comment]
@daniyar-niantic I plotted the six-dof information for gt vs what a well-trained PoseCNN is predicts. The PoseCNN is making uniformly distributed predictions for all six-dof parameters, while the GT has a small number of significant outliers, especially for tx and ty. It does seem that there is considerably more noise in the computed poses relative to what a well-trained PoseCNN predicts. So that does explain it.

I wanted the PoseCNN to learn the velo2velo transformation because it seems like the GT poses are too noisy for the Depth Model to learn accurate metric depth. I thought that if I sandwich the pose that the PoseCNN is learning between the cam2velo and velo2cam transformations, then it would be forced to learn everything to metric scale because the cam2velo and velo2cam transformations have the metric information encoded in them. And because we are still using the PoseCNN, we no longer have to worry about noisy measurements affecting our depth estimates. But it didn't seem to do that. The median scaling ratios are consistently 19, but their range is 16, suggesting that the scaling ratio the model is learning is still not consistent. When I trained the DepthCNN with GT pose, the scaling ratios were consistently 1.0 with a range of less than 1.0. However, this is still far better than having the PoseCNN learn the entire cam2cam transformation. Doing it that way, the scaling ratios are all over the place. I just don't understand why the scaling ratio isn't 1.0 when learning the velo2velo transformation or why it doesn't at least have a consistent scaling ratio.

**Comment by aj96** [comment]
@daniyar-niantic Actually, nvm. I get it. You are actually not applying a constraint when you learn the velo2velo transformation. You can still scale the depth by k and the translation components by 1/k such that you still get the same re-projection. Only if you set the cam2cam transformation so that no aspect of its translations can be learned, will it learn metric depth. It's still curious though why the scaling ratios are so much more consistent...

**Comment by YJonmo** [comment]
I am also trying to get the pose network trained using the GT poses (supervised learning). It just doesn't converge.
